### DISCUSSION TITLE
example code error: access is denied

### DISCUSSION THREAD

**Discussion by impromptu1583** [hbldh/bleak#985] [Q&A] [answered]
I'm copied the example script enable_notifications.py and only changed the ADDRESS and CHARACTERISTIC_UUID values (which I obtained using the server_explorer.py example) and I get this error:

```
Traceback (most recent call last):
  File "C:\Users\mcmil\Documents\py\btest.py", line 43, in <module>
    asyncio.run(
  File "C:\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Python310\lib\asyncio\base_events.py", line 646, in run_until_complete
    return future.result()
  File "C:\Users\mcmil\Documents\py\btest.py", line 37, in main
    await client.start_notify(char_uuid, notification_handler)
  File "C:\Python310\lib\site-packages\bleak\backends\winrt\client.py", line 787, in start_notify
    await characteristic_obj.write_client_characteristic_configuration_descriptor_async(
PermissionError: [WinError -2147024891] Access is denied
```

What is wrong? Something with the device or something with my computer / install / permissions?

Windows 11, python 3.10, bleak0.16.0

(for context the device I'm connecting to is an arduino mkr 1010 running the simple BLE example from [this tutorial](https://docs.arduino.cc/tutorials/mkr-wifi-1010/enabling-ble))

**Comment by dlech** [comment, marked as answer]
Does this characteristic actually support notifications (this should be able to be seen in the properties listed by the service explorer example program).

Does this characteristic require authentication? If so, the device must be paired first.

**Reply by impromptu1583** [reply]
Thank you for your reply!

Yes, here's the line from the service explorer script:
`INFO:__main__:  [Characteristic] 0c9f3913-2a36-4380-be3a-5c81c933f0c6 (Handle: 13): Unknown (read,notify), Value: b'X'`
As you can see it successfully read a value as well.

I can also connect to the device successfully using the LightBlue android app which also detects the characteristic as having "notify" support and can successfully "subscribe" to notifications.

**Reply by dlech** [reply]
You could try logging Bluetooth packets with Wireshark (see troubleshooting page in Bleak docs) to see if the request is actually going out over the air and what type of reply is received from the device.

**Reply by dlech** [reply]
Possibly the same issue? https://answers.microsoft.com/en-us/windows/forum/all/bluetooth-low-energy/364c65d0-17bf-446d-b635-31d4a08eccee

**Reply by dlech** [reply]
Can you also share the output of the service explorer? Could be something like https://stackoverflow.com/a/69126661/1976323.

**Reply by impromptu1583** [reply]
```
INFO:bleak.backends.winrt.client:Services resolved for BleakClientWinRT (30:C6:F7:02:99:E6)
INFO:__main__:Connected: True
INFO:__main__:[Service] 00001800-0000-1000-8000-00805f9b34fb (Handle: 1): Generic Access Profile
INFO:__main__:  [Characteristic] 00002a00-0000-1000-8000-00805f9b34fb (Handle: 2): Device Name (read), Value: b'Arduino'
INFO:__main__:  [Characteristic] 00002a01-0000-1000-8000-00805f9b34fb (Handle: 4): Appearance (read), Value: b'\x00\x00'
INFO:__main__:[Service] 00001801-0000-1000-8000-00805f9b34fb (Handle: 6): Generic Attribute Profile
INFO:__main__:  [Characteristic] 00002a05-0000-1000-8000-00805f9b34fb (Handle: 7): Service Changed (indicate), Value: None
INFO:__main__:          [Descriptor] 00002902-0000-1000-8000-00805f9b34fb (Handle: 9): Client Characteristic Configuration) | Value: b'\x02\x00'
INFO:__main__:[Service] 0000180a-0000-1000-8000-00805f9b34fb (Handle: 10): Device Information
INFO:__main__:  [Characteristic] 00002a57-0000-1000-8000-00805f9b34fb (Handle: 11): Digital Output (read,write), Value: b'\x00'
INFO:__main__:  [Characteristic] 00002a58-0000-1000-8000-00805f9b34fb (Handle: 13): Analog (read,notify), Value: b'\xc8'
INFO:__main__:          [Descriptor] 00002902-0000-1000-8000-00805f9b34fb (Handle: 15): Client Characteristic Configuration) | Value: b'\x00\x00'
```

seems to have the correct 2902 descriptor

**Reply by dlech** [reply]
The characteristic in question is included in the Device Information service though. These are not valid characteristics for that service. If you control the firmware, you could try using a custom service instead of the Device Information service.

**Reply by impromptu1583** [reply]
I tried changing the service to 1801 Generic Attribute Profile as well as using a random custom UUID for it - both with the same result.

```
INFO:bleak.backends.winrt.client:Services resolved for BleakClientWinRT (30:C6:F7:02:99:E6)
INFO:__main__:Connected: True
INFO:__main__:[Service] 00001800-0000-1000-8000-00805f9b34fb (Handle: 1): Generic Access Profile
INFO:__main__:  [Characteristic] 00002a00-0000-1000-8000-00805f9b34fb (Handle: 2): Device Name (read), Value: b'Arduino'
INFO:__main__:  [Characteristic] 00002a01-0000-1000-8000-00805f9b34fb (Handle: 4): Appearance (read), Value: b'\x00\x00'
INFO:__main__:[Service] 00001801-0000-1000-8000-00805f9b34fb (Handle: 6): Generic Attribute Profile
INFO:__main__:  [Characteristic] 00002a05-0000-1000-8000-00805f9b34fb (Handle: 7): Service Changed (indicate), Value: None
INFO:__main__:          [Descriptor] 00002902-0000-1000-8000-00805f9b34fb (Handle: 9): Client Characteristic Configuration) | Value: b'\x02\x00'
INFO:__main__:[Service] 0c9f3913-2a36-4380-be3a-5c81c933f0c6 (Handle: 10): Unknown
INFO:__main__:  [Characteristic] 00002a58-0000-1000-8000-00805f9b34fb (Handle: 11): Analog (read,notify), Value: b'\xc8'
INFO:__main__:          [Descriptor] 00002902-0000-1000-8000-00805f9b34fb (Handle: 13): Client Characteristic Configuration) | Value: b'\x00\x00'
```

and the notify script:

```
  File "C:\Python310\lib\site-packages\bleak\backends\winrt\client.py", line 787, in start_notify
    await characteristic_obj.write_client_characteristic_configuration_descriptor_async(
PermissionError: [WinError -2147024891] Access is denied
```

**Reply by dlech** [reply]
Did you try logging Bluetooth packets?

**Reply by impromptu1583** [reply]
Actually I was away from my desk for a bit and came back and now it works. Not 100% sure, but I think changing the service UUID fixed it and I just needed to wait for the bluetooth cache to clear on windows. Thank you for your help!
